### PR TITLE
Stock Service feature

### PR DESCRIPTION
**Pull request by JuanGQCadavid**

### CODE REVIEW

File: stock-service/src/main/java/com/ecommercepractice/stockservice/StockServiceApplication.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+public class StockServiceApplication {
+
+	public static void main(String[] args) {
+		SpringApplication.run(StockServiceApplication.class, args);
+	}
+
```
Comment (czuluagah) [inline stock-service/src/main/java/com/ecommercepractice/stockservice/StockServiceApplication.java]: White Spaces!!!

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    StockService stockService;
+
+    @Autowired
+    StockModelAssembler stockModelAssembler;
+
+    @ApiOperation(value = "APPEND_TO_STOCK", notes = "Append a product to stock")
+    @PostMapping()
+    public ResponseEntity<EntityModel<Stock>> appendToInventory(@Valid @RequestBody Stock productStock) {
+        log.info(String.format("STOCK | APPEND | PAYLOAD { Stock -> %s } ", productStock));
+        return new ResponseEntity(stockModelAssembler.toModel(stockService.appendToInventory(productStock)), HttpStatus.CREATED);
```
Comment (czuluagah) [inline stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java]: Reformat this line

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @ApiOperation(value ="FILTER", notes = "Fetch products by filtering them by price, stockQuantity and soldQuiantity" +
+            "Also, you can specify whether you want to list not onlyProducts also")
+    @GetMapping()
+    public ResponseEntity<List<EntityModel<Stock>>>fetchByFiltering(@RequestParam(required = false) Float price,
```
Comment (czuluagah) [inline stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java]: add space bettween method name and Return type

Comment (czuluagah) [reply stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java]: Try to use @BeanParam def

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        List<EntityModel<Stock>> stockList = stockService.fetchByFiltering(price, stockQuantity, soldQuantity, onlyActive)
+                .stream()
+                .map(stockModelAssembler::toModel)
+                .collect(Collectors.toList());
+
+        return new ResponseEntity(stockList, HttpStatus.OK);
+    }
+
+    @ApiOperation(value ="FETCH BY ID", notes = "Get a product on stock by their id.")
+    @GetMapping("/{stockId}")
```
Comment (czuluagah) [inline stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java]: please, try to set a resource more clear for example /stock/products/{product-id}

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    public ResponseEntity<EntityModel<Stock>> fetchByStockId(@PathVariable(required = true) Long stockId) {
+        log.info(String.format("STOCK | FETCH | PAYLOAD { StockId -> %s } ", stockId));
+        return new ResponseEntity(stockModelAssembler.toModel(stockService.fetchByStockId(stockId)), HttpStatus.OK);
+    }
+
+    @ApiOperation(value ="UPDATE BY ID", notes = "Update stockQuantity and/or salePrice on stock by their id.")
+    @PutMapping("/{stockId}")
+    public ResponseEntity<EntityModel<Stock>> updateByStockId(@PathVariable(required = true) Long stockId,
```
Comment (czuluagah) [inline stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java:73]: please, try to set a resource more clear for example /stock/products/{product-id}

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                                                 @RequestParam(required = false) Float salePrice
+    ) {
+        log.info(String.format("STOCK | UPDATE | PAYLOAD { StockId -> %s , Stock -> %s } ", stockId,
+                stockQuantity == null ? "NO_QUANTITY" : stockQuantity,
+                salePrice == null ? "NO_SALE_PRICE" : salePrice
+        ));
+        return new ResponseEntity(stockModelAssembler.toModel(stockService.updateByStockId(stockId, stockQuantity, salePrice)), HttpStatus.OK);
+    }
+    @ApiOperation(value ="DELETE BY ID", notes = "DELETE a product on stock by their id.")
+    @DeleteMapping("/{stockId}")
```
Comment (czuluagah) [inline stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java]: please, try to set a resource more clear for example /stock/products/{product-id}

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    public ResponseEntity deleteByStockId(@PathVariable(required = true) Long stockId) {
+        log.info(String.format("STOCK | DELETE | PAYLOAD { StockId -> %s } ", stockId));
+        stockService.deleteByStockId(stockId);
+        return new ResponseEntity(HttpStatus.NO_CONTENT);
+    }
+
+    @ApiOperation(value ="FREEZE BY ID", notes = "FREEZE a product on stock by their id. When a product is freezed means" +
+            "that is inside a car.")
+    @PostMapping("/freeze/{stockId}")
```
Comment (czuluagah) [inline stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java]: stock/products/{id-product}/freeze/{quantity}

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    public ResponseEntity freezeStockItem(@PathVariable(required = true) Long stockId, @RequestParam(required = true) Long freezeQuantities) {
+        log.info(String.format("STOCK | FREEZE | PAYLOAD { StockId -> %s} ", stockId));
+        stockService.freezeStockItem(stockId, freezeQuantities);
+        return new ResponseEntity(HttpStatus.NO_CONTENT);
+    }
+
+    @ApiOperation(value ="BOUGHT BY ID", notes = "BOUGHT a product on stock by their id. That means, the product is" +
+            "already paid, and it is not available for being purchased")
+    @PostMapping("/bought/{stockId}")
+    public ResponseEntity boughtStockItem(@PathVariable(required = true) Long stockId, @RequestParam Long boughtQuantity) {
```
Comment (czuluagah) [inline stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java]: same, please use a mapping more descriptive

File: stock-service/src/main/java/com/ecommercepractice/stockservice/StockServiceApplication.java (hunk 1 above)
Comment (JuanGQCadavid) [reply stock-service/src/main/java/com/ecommercepractice/stockservice/StockServiceApplication.java]: Done

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java (hunk 2 above)
Comment (JuanGQCadavid) [reply stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java]: Done

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java (hunk 3 above)
Comment (JuanGQCadavid) [reply stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java]: > add space bettween method name and Return type

Done

Comment (JuanGQCadavid) [reply stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java]: > Try to use @BeanParam def

Done

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java (hunk 4 above)
Comment (JuanGQCadavid) [reply stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java]: Done

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java (hunk 5 above)
Comment (JuanGQCadavid) [reply stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java:73]: done

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java (hunk 6 above)
Comment (JuanGQCadavid) [reply stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java]: done

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java (hunk 7 above)
Comment (JuanGQCadavid) [reply stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java]: done

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java (hunk 8 above)
Comment (JuanGQCadavid) [reply stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java]: done

File: stock-service/src/main/resources/application.properties [hunk 9]
```diff
@@ -0,0 +1 @@
+
```
Comment (czuluagah) [inline stock-service/src/main/resources/application.properties]: use this property to identify the app name
#spring.application.name

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        return new ResponseEntity(HttpStatus.NO_CONTENT);
+    }
+
+    @ApiOperation(value ="BOUGHT BY ID", notes = "BOUGHT a product on stock by their id. That means, the product is" +
+            "already paid, and it is not available for being purchased")
+    @PostMapping("products/{stockId}/buy/{QuantityToBuy}")
+    public ResponseEntity boughtStockItem(@PathVariable(required = true) Long stockId, @PathVariable Long QuantityToBuy) {
+        log.info(String.format("STOCK | BOUGHT | PAYLOAD { StockId -> %s } ", stockId));
+        stockService.boughtStockItem(stockId, QuantityToBuy);
+        return new ResponseEntity(HttpStatus.NO_CONTENT);
```
Comment (czuluagah) [inline stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java:106]: If this method does not return a body try to change to Response only

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        stockService.deleteByStockId(stockId);
+        return new ResponseEntity(HttpStatus.NO_CONTENT);
+    }
+
+    @ApiOperation(value ="FREEZE BY ID", notes = "FREEZE a product on stock by their id. When a product is freezed means" +
+            "that is inside a car.")
+    @PostMapping("products/{stockId}/freeze/{freezeQuantities}")
+    public ResponseEntity freezeStockItem(@PathVariable(required = true) Long stockId, @PathVariable(required = true) Long freezeQuantities) {
+        log.info(String.format("STOCK | FREEZE | PAYLOAD { StockId -> %s} ", stockId));
+        stockService.freezeStockItem(stockId, freezeQuantities);
```
Comment (czuluagah) [inline stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java:96]: If this method does not return a body try to change to Response only

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+                stockQuantity == null ? "NO_QUANTITY" : stockQuantity,
+                salePrice == null ? "NO_SALE_PRICE" : salePrice
+        ));
+        return new ResponseEntity(stockModelAssembler.toModel(stockService.updateByStockId(stockId, stockQuantity, salePrice)), HttpStatus.OK);
+    }
+    @ApiOperation(value ="DELETE BY ID", notes = "DELETE a product on stock by their id.")
+    @DeleteMapping("products/{stockId}")
+    public ResponseEntity deleteByStockId(@PathVariable(required = true) Long stockId) {
+        log.info(String.format("STOCK | DELETE | PAYLOAD { StockId -> %s } ", stockId));
+        stockService.deleteByStockId(stockId);
```
Comment (czuluagah) [inline stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java:87]: If this method does not return a body try to change to Response only

File: stock-service/src/main/resources/application.properties (hunk 9 above)
Comment (JuanGQCadavid) [reply stock-service/src/main/resources/application.properties]: Done

File: stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java (hunk 10 above)
Comment (JuanGQCadavid) [reply stock-service/src/main/java/com/ecommercepractice/stockservice/controller/StockController.java:106]: I don't understand, what kind of object should the method respond? I only found Response in Java ws rs core package I don't consider that appropriate.

**Pull request merged**